File: Google/CodeJam/2019/qualification/C/c.go
package main

import (
	"fmt"
	"math/big"
	"sort"
)

func gcd(a, b *big.Int) *big.Int {
	if b.Cmp(big.NewInt(0)) == 0 {
		return a
	}
	return gcd(b, a.Mod(a, b))
}

type bigIntArray []*big.Int

func (s bigIntArray) Len() int {
	return len(s)
}

func (s bigIntArray) Swap(i, j int) {
	s[i], s[j] = s[j], s[i]
}

func (s bigIntArray) Less(i, j int) bool {
	return s[i].Cmp(s[j]) <= 0 // x < y
}

func solve() string {
	ans := ""

	n := new(big.Int)
	fmt.Scanf("%s", n)
	l := 0
	fmt.Scanf("%d", &l)

	inp := make([]*big.Int, l)
	for i := 0; i < l; i++ {
		inp[i] = new(big.Int)
		fmt.Scanf("%d", inp[i])
		// log.Println("inp", i, inp[i])
	}

	res := make([]*big.Int, l+1)
	for i := 0; i < l+1; i++ {
		res[i] = new(big.Int)
	}

	idx := -1
	for i := 1; i < l; i++ {
		if inp[i].Cmp(inp[i-1]) == 0 { // AAA or ABA
			continue
		}

		a := new(big.Int)
		a.SetString(inp[i-1].String(), 10)
		b := new(big.Int)
		b.SetString(inp[i].String(), 10)
		g := gcd(a, b)
		idx = i
		res[i] = g
		break
	}
	// if idx == -1 {
	// 	log.Fatalln("Fucked")
	// }
	// log.Println(idx, res[idx])

	s := make(map[string]bool)
	for i := idx - 1; i >= 0; i-- {
		// log.Println(inp[i], res[i+1])
		res[i].Div(inp[i], res[i+1])
		// log.Println(i, res[i])
	}
	for i := idx + 1; i < l+1; i++ {
		res[i].Div(inp[i-1], res[i-1])
	}
	for i := 0; i < l+1; i++ {
		s[res[i].String()] = true
	}

	primeString := make(bigIntArray, 0)
	for key := range s {
		tmp := new(big.Int)
		fmt.Sscanf(key, "%s", tmp)
		primeString = append(primeString, tmp)
	}
	sort.Sort(primeString)

	prime := make(map[string]string)
	for i, str := range primeString {
		prime[str.String()] = string(i + 'A')
	}

	for i := 0; i < l+1; i++ {
		ans += prime[res[i].String()]
	}

	return ans
}

func main() {
	ncase := 0
	fmt.Scanf("%d", &ncase)
	for i := 1; i <= ncase; i++ {
		res := solve()
		fmt.Printf("Case #%d: %s\n", i, res)
	}
}

File: Google/CodeJam/2019/qualification/c.cpp
#include <bits/stdc++.h>

using namespace std;

typedef pair<int, int> ii;

int gcd(int a, int b)
{
    a = abs(a);
    b = abs(b);
    // return a == 0 ? b : gcd(b % a, a);
    return b == 0 ? a : gcd(b, a % b);
}

string solve()
{
    string ans;

    int n, l;
    scanf("%d %d", &n, &l);

    int inp[l];
    for (int i = 0; i < l; i++)
        scanf("%d", &inp[i]);

    int res[l + 1];
    int idx = -1;
    for (int i = 1; i < l; i++) {
        if (inp[i] == inp[i - 1]) // AAA or ABA
            continue;

        int g = gcd(inp[i - 1], inp[i]);
        idx = i;
        res[i] = g;
    }

    set<int> s;
    for (int i = idx - 1; i >= 0; i--)
        res[i] = inp[i] / res[i + 1];
    for (int i = idx + 1; i < l + 1; i++)
        res[i] = inp[i - 1] / res[i - 1];
    for (int i = 0; i < l + 1; i++)
        s.insert(res[i]);

    map<int, char> prime;
    assert(26 == s.size());
    auto it = s.begin();
    for (int i = 0; i < 26; i++) {
        prime[*it] = i + 'A';
        it++;
    }

    for (int i = 0; i < l + 1; i++) {
        ans += prime[res[i]];
    }

    return ans;
}

int main()
{
    int ncase;
    scanf("%d", &ncase);
    for (int i = 1; i <= ncase; i++) {
        auto res = solve();
        printf("Case #%d: %s\n", i, res.c_str());
    }
}
